cas-server-core-services/src/test/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDaoTests.java

```diff
@@ -0,0 +1,391 @@
+
+        final RegisteredService r2 = this.dao.save(r);
+        final RegisteredService r3 = this.dao.findServiceById(r2.getId());
+
+        assertEquals(r, r2);
+        assertEquals(r2, r3);
+        assertNotNull(r3.getAttributeReleasePolicy());
+        assertEquals(r2.getAttributeReleasePolicy(), r3.getAttributeReleasePolicy());
+    }
+
+    @Test
+    public void verifySaveAttributeReleasePolicyAllowedAttrRulesAndFilter() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setName("testSaveAttributeReleasePolicyAllowedAttrRulesAndFilter");
+        r.setServiceId("testId");
+        r.setTheme("testtheme");
+        r.setEvaluationOrder(1000);
+        r.setAccessStrategy(new DefaultRegisteredServiceAccessStrategy(true, false));
+        r.setProxyPolicy(new RegexMatchingRegisteredServiceProxyPolicy("https://.+"));
+        r.setRequiredHandlers(new HashSet<>(Arrays.asList("h1", "h2")));
+
+        final ReturnAllowedAttributeReleasePolicy policy = new ReturnAllowedAttributeReleasePolicy();
+        policy.setAllowedAttributes(Arrays.asList("1", "2", "3"));
+        r.setAttributeReleasePolicy(policy);
+        r.getAttributeReleasePolicy().setAttributeFilter(new RegisteredServiceRegexAttributeFilter("\\w+"));
+
+        final RegisteredService r2 = this.dao.save(r);
+        final RegisteredService r3 = this.dao.findServiceById(r2.getId());
+
+        assertEquals(r, r2);
+        assertEquals(r2, r3);
+        assertNotNull(r3.getAttributeReleasePolicy());
+        assertEquals(r2.getAttributeReleasePolicy(), r3.getAttributeReleasePolicy());
+    }
+
+    @Test
+    public void verifyServiceType() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setServiceId("^https://.+");
+        r.setName("testServiceType");
+        r.setTheme("testtheme");
+        r.setEvaluationOrder(1000);
+
+        final RegisteredService r2 = this.dao.save(r);
+        assertTrue(r2 instanceof  RegexRegisteredService);
+    }
+
+    @Test(expected=RuntimeException.class)
+    public void verifyServiceWithInvalidFileName() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setServiceId("^https://.+");
+        r.setName("hell/o@world:*");
+        r.setEvaluationOrder(1000);
+
+        final RegisteredService r2 = this.dao.save(r);
+    }
+    
+    @Test
+    public void verifyServiceRemovals() throws Exception{
+        final List<RegisteredService> list = new ArrayList<>(5);
+        IntStream.range(1, 5).forEach(i -> {
+            final RegexRegisteredService r = new RegexRegisteredService();
+            r.setServiceId("^https://.+");
+            r.setName("testServiceType");
+            r.setTheme("testtheme");
+            r.setEvaluationOrder(1000);
+            r.setId(i * 100);
+            list.add(this.dao.save(r));
+        });
+
+        list.stream().forEach(r2 -> {
+            try {
+                Thread.sleep(500);
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: ['500' is a magic number.](https://www.codacy.com/app/mmoayyed/cas/file/2643620500/issues/source?bid=3256115&fileBranchId=3322154#l270)